### PR TITLE
Dobble example

### PR DESCRIPTION
**Pull request by krtab**
I tried encoding the generation of [Dobble](https://en.wikipedia.org/wiki/Dobble) cards in owi/Z3.
The performances so far are abysmal but I'm opening a draft PR if we want to refer to it later.

### CODE REVIEW

**Comment by krtab** [comment]
I tried implementing it directly in z3 (https://github.com/krtab/dobble_z3/blob/master/src/main.rs) and to my surprise it is much faster. (Takes only around 200ms while the owi exec takes ~30s). I checked if we were inefficient in owi but 96% of the time is spent in z3.
I think it would be interesting to check one happens. One thing I suspect is that too much branching is discharged to the solver, even when the for loop are only checking concrete values.

**Comment by filipeom** [comment]
> I tried implementing it directly in z3 (https://github.com/krtab/dobble_z3/blob/master/src/main.rs) and to my surprise it is much faster. (Takes only around 200ms while the owi exec takes ~30s). I checked if we were inefficient in owi but 96% of the time is spent in z3. I think it would be interesting to check one happens. One thing I suspect is that too much branching is discharged to the solver, even when the for loop are only checking concrete values.

I think the comparison with the direct Z3 encoding is not really fair as you're testing 1 satisfiability check with possibly many during wasm symbolic execution. However, I also find strange why this takes so long because this shouldn't generate to much paths.

And, even if you're discharging many concrete queries to the solver, Z3 is usually fast resolving these. Incurring only a considerable overhead.

I reckon hc might be playing a major factor in slowing this program down. If you want you can try compiling with maximal optimizations to only use local variables:

Applying this diff:

```diff
diff --git a/src/cmd_c.ml b/src/cmd_c.ml
index e6ca937..70585f6 100644
--- a/src/cmd_c.ml
+++ b/src/cmd_c.ml
@@ -18,10 +18,10 @@ type metadata =
 let clang bin ~flags ~out file : Bos.Cmd.t =
   Cmd.(bin %% flags % "-o" % p out % p file)

-let opt bin file : Bos.Cmd.t = Cmd.(bin % "-O1" % "-o" % p file % p file)
+let opt bin file : Bos.Cmd.t = Cmd.(bin % "-O3" % "-o" % p file % p file)

 let llc bin ~bc ~obj : Bos.Cmd.t =
-  let flags = Cmd.of_list [ "-O1"; "-march=wasm32"; "-filetype=obj"; "-o" ] in
+  let flags = Cmd.of_list [ "-O3"; "-march=wasm32"; "-filetype=obj"; "-o" ] in
   Cmd.(bin %% flags % p obj % p bc)

 let ld bin ~flags ~out files : Bos.Cmd.t =
```

And running `owi -c -O3`

I already did this, but it needs `i32.popcnt`:

```wast
(module
  (type (;0;) (func (result i32)))
  (type (;1;) (func (param i32)))
  (type (;2;) (func))
  (import "symbolic" "i32_symbol" (func $__i32 (type 0)))
  (import "symbolic" "assume" (func $__owi_assume (type 1)))
  (import "symbolic" "assert" (func $__owi_assert (type 1)))
  (func $__original_main (type 0) (result i32)
    (local i32 i32 i32 i32 i32 i32 i32 i32 i32 i32)
    call $owi_i32
    local.tee 0
    i32.const 128
    i32.lt_u
    call $owi_assume
    local.get 0
    i32.const 1
    i32.shr_u
    i32.const 1
    i32.and
    local.get 0
    i32.const 1
    i32.and
    i32.add
    local.get 0
    i32.const 2
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 0
    i32.const 3
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 0
    i32.const 4
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 0
    i32.const 5
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 0
    i32.const 6
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    i32.const 3
    i32.eq
    call $owi_assume
    call $owi_i32
    local.tee 1
    i32.const 128
    i32.lt_u
    call $owi_assume
    local.get 1
    i32.const 1
    i32.shr_u
    i32.const 1
    i32.and
    local.get 1
    i32.const 1
    i32.and
    i32.add
    local.get 1
    i32.const 2
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 1
    i32.const 3
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 1
    i32.const 4
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 1
    i32.const 5
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 1
    i32.const 6
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    i32.const 3
    i32.eq
    call $owi_assume
    call $owi_i32
    local.tee 2
    i32.const 128
    i32.lt_u
    call $owi_assume
    local.get 2
    i32.const 1
    i32.shr_u
    i32.const 1
    i32.and
    local.get 2
    i32.const 1
    i32.and
    i32.add
    local.get 2
    i32.const 2
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 2
    i32.const 3
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 2
    i32.const 4
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 2
    i32.const 5
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 2
    i32.const 6
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    i32.const 3
    i32.eq
    call $owi_assume
    call $owi_i32
    local.tee 3
    i32.const 128
    i32.lt_u
    call $owi_assume
    local.get 3
    i32.const 1
    i32.shr_u
    i32.const 1
    i32.and
    local.get 3
    i32.const 1
    i32.and
    i32.add
    local.get 3
    i32.const 2
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 3
    i32.const 3
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 3
    i32.const 4
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 3
    i32.const 5
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 3
    i32.const 6
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    i32.const 3
    i32.eq
    call $owi_assume
    call $owi_i32
    local.tee 4
    i32.const 128
    i32.lt_u
    call $owi_assume
    local.get 4
    i32.const 1
    i32.shr_u
    i32.const 1
    i32.and
    local.get 4
    i32.const 1
    i32.and
    i32.add
    local.get 4
    i32.const 2
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 4
    i32.const 3
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 4
    i32.const 4
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 4
    i32.const 5
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 4
    i32.const 6
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    i32.const 3
    i32.eq
    call $owi_assume
    call $owi_i32
    local.tee 5
    i32.const 128
    i32.lt_u
    call $owi_assume
    local.get 5
    i32.const 1
    i32.shr_u
    i32.const 1
    i32.and
    local.get 5
    i32.const 1
    i32.and
    i32.add
    local.get 5
    i32.const 2
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 5
    i32.const 3
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 5
    i32.const 4
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 5
    i32.const 5
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 5
    i32.const 6
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    i32.const 3
    i32.eq
    call $owi_assume
    call $owi_i32
    local.tee 6
    i32.const 128
    i32.lt_u
    call $owi_assume
    local.get 6
    i32.const 1
    i32.shr_u
    i32.const 1
    i32.and
    local.get 6
    i32.const 1
    i32.and
    i32.add
    local.get 6
    i32.const 2
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 6
    i32.const 3
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 6
    i32.const 4
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 6
    i32.const 5
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    local.get 6
    i32.const 6
    i32.shr_u
    i32.const 1
    i32.and
    i32.add
    i32.const 3
    i32.eq
    call $owi_assume
    local.get 0
    local.get 1
    i32.ne
    call $owi_assume
    local.get 0
    local.get 2
    i32.ne
    call $owi_assume
    i32.const 0
    local.set 7
    i32.const 0
    local.set 8
    block  ;; label = @1
      local.get 1
      local.get 0
      i32.and
      i32.popcnt
      i32.const 1
      i32.ne
      br_if 0 (;@1;)
      i32.const 0
      local.set 8
      local.get 2
      local.get 0
      i32.and
      local.tee 9
      i32.eqz
      br_if 0 (;@1;)
      local.get 9
      local.get 9
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 8
    end
    local.get 0
    local.get 3
    i32.ne
    call $owi_assume
    block  ;; label = @1
      local.get 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 3
      local.get 0
      i32.and
      local.tee 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 8
      local.get 8
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 7
    end
    local.get 0
    local.get 4
    i32.ne
    call $owi_assume
    i32.const 0
    local.set 9
    i32.const 0
    local.set 8
    block  ;; label = @1
      local.get 7
      i32.eqz
      br_if 0 (;@1;)
      i32.const 0
      local.set 8
      local.get 4
      local.get 0
      i32.and
      local.tee 7
      i32.eqz
      br_if 0 (;@1;)
      local.get 7
      local.get 7
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 8
    end
    local.get 0
    local.get 5
    i32.ne
    call $owi_assume
    block  ;; label = @1
      local.get 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 5
      local.get 0
      i32.and
      local.tee 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 8
      local.get 8
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 9
    end
    local.get 0
    local.get 6
    i32.ne
    call $owi_assume
    i32.const 0
    local.set 7
    i32.const 0
    local.set 8
    block  ;; label = @1
      local.get 9
      i32.eqz
      br_if 0 (;@1;)
      i32.const 0
      local.set 8
      local.get 6
      local.get 0
      i32.and
      local.tee 0
      i32.eqz
      br_if 0 (;@1;)
      local.get 0
      local.get 0
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 8
    end
    local.get 1
    local.get 2
    i32.ne
    call $owi_assume
    block  ;; label = @1
      local.get 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 2
      local.get 1
      i32.and
      local.tee 0
      i32.eqz
      br_if 0 (;@1;)
      local.get 0
      local.get 0
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 7
    end
    local.get 1
    local.get 3
    i32.ne
    call $owi_assume
    i32.const 0
    local.set 8
    i32.const 0
    local.set 0
    block  ;; label = @1
      local.get 7
      i32.eqz
      br_if 0 (;@1;)
      i32.const 0
      local.set 0
      local.get 3
      local.get 1
      i32.and
      local.tee 7
      i32.eqz
      br_if 0 (;@1;)
      local.get 7
      local.get 7
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 0
    end
    local.get 1
    local.get 4
    i32.ne
    call $owi_assume
    block  ;; label = @1
      local.get 0
      i32.eqz
      br_if 0 (;@1;)
      local.get 4
      local.get 1
      i32.and
      local.tee 0
      i32.eqz
      br_if 0 (;@1;)
      local.get 0
      local.get 0
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 8
    end
    local.get 1
    local.get 5
    i32.ne
    call $owi_assume
    i32.const 0
    local.set 7
    i32.const 0
    local.set 0
    block  ;; label = @1
      local.get 8
      i32.eqz
      br_if 0 (;@1;)
      i32.const 0
      local.set 0
      local.get 5
      local.get 1
      i32.and
      local.tee 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 8
      local.get 8
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 0
    end
    local.get 1
    local.get 6
    i32.ne
    call $owi_assume
    local.get 2
    local.get 3
    i32.ne
    call $owi_assume
    block  ;; label = @1
      local.get 0
      i32.eqz
      br_if 0 (;@1;)
      local.get 6
      local.get 1
      i32.and
      local.tee 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 1
      i32.popcnt
      i32.const 1
      i32.gt_u
      br_if 0 (;@1;)
      local.get 3
      local.get 2
      i32.and
      local.tee 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 1
      local.get 1
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 7
    end
    local.get 2
    local.get 4
    i32.ne
    call $owi_assume
    i32.const 0
    local.set 0
    i32.const 0
    local.set 1
    block  ;; label = @1
      local.get 7
      i32.eqz
      br_if 0 (;@1;)
      i32.const 0
      local.set 1
      local.get 4
      local.get 2
      i32.and
      local.tee 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 8
      local.get 8
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 1
    end
    local.get 2
    local.get 5
    i32.ne
    call $owi_assume
    block  ;; label = @1
      local.get 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 5
      local.get 2
      i32.and
      local.tee 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 1
      local.get 1
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 0
    end
    local.get 2
    local.get 6
    i32.ne
    call $owi_assume
    local.get 3
    local.get 4
    i32.ne
    call $owi_assume
    i32.const 0
    local.set 8
    i32.const 0
    local.set 1
    block  ;; label = @1
      local.get 0
      i32.eqz
      br_if 0 (;@1;)
      i32.const 0
      local.set 1
      local.get 6
      local.get 2
      i32.and
      local.tee 0
      i32.eqz
      br_if 0 (;@1;)
      i32.const 0
      local.set 1
      local.get 0
      i32.popcnt
      i32.const 1
      i32.gt_u
      br_if 0 (;@1;)
      i32.const 0
      local.set 1
      local.get 4
      local.get 3
      i32.and
      local.tee 0
      i32.eqz
      br_if 0 (;@1;)
      local.get 0
      local.get 0
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 1
    end
    local.get 3
    local.get 5
    i32.ne
    call $owi_assume
    block  ;; label = @1
      local.get 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 5
      local.get 3
      i32.and
      local.tee 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 1
      local.get 1
      i32.const -1
      i32.add
      i32.and
      i32.eqz
      local.set 8
    end
    local.get 3
    local.get 6
    i32.ne
    call $owi_assume
    local.get 4
    local.get 5
    i32.ne
    call $owi_assume
    i32.const 1
    local.set 0
    block  ;; label = @1
      local.get 8
      i32.eqz
      br_if 0 (;@1;)
      local.get 6
      local.get 3
      i32.and
      local.tee 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 1
      i32.popcnt
      i32.const 1
      i32.gt_u
      br_if 0 (;@1;)
      local.get 5
      local.get 4
      i32.and
      local.tee 1
      i32.eqz
      br_if 0 (;@1;)
      local.get 1
      local.get 1
      i32.const -1
      i32.add
      i32.and
      i32.const 0
      i32.ne
      local.set 0
    end
    local.get 4
    local.get 6
    i32.ne
    call $owi_assume
    local.get 5
    local.get 6
    i32.ne
    call $owi_assume
    local.get 0
    local.get 6
    local.get 4
    i32.and
    local.tee 1
    i32.eqz
    i32.or
    local.get 1
    local.get 1
    i32.const -1
    i32.add
    i32.and
    i32.const 0
    i32.ne
    i32.or
    local.get 6
    local.get 5
    i32.and
    local.tee 6
    i32.eqz
    i32.or
    local.get 6
    local.get 6
    i32.const -1
    i32.add
    i32.and
    i32.const 0
    i32.ne
    i32.or
    call $owi_assert
    i32.const 0)
  (func $owi_i32 (type 0) (result i32)
    call $__i32)
  (func $owi_assume (type 1) (param i32)
    local.get 0
    call $__owi_assume)
  (func $owi_assert (type 1) (param i32)
    local.get 0
    call $__owi_assert)
  (func $_start (type 2)
    call $__original_main
    drop)
  (table (;0;) 1 1 funcref)
  (memory (;0;) 129)
  (global $__stack_pointer (mut i32) (i32.const 8389632))
  (export "memory" (memory 0))
  (export "_start" (func $_start)))
```

**Comment by filipeom** [comment]
Nevermind, you are right. I'm also stuck in the solver idk why. I'm printing the satisfiability checks and they all seem very trivial to me

**Comment by krtab** [comment]
Hi @filipeom we just solved it with @zapashcanon  and @chambart ! It boils down to the `&&` in C are creating a lot of branches. We have discussed ways to handle that as well but nothing conclusive, I'll update the PR.

**Comment by filipeom** [comment]
> Hi @filipeom we just solved it with @zapashcanon and @chambart ! It boils down to the lazy `&&` in C are creating a lot of branches. We have discussed ways to handle that as well but nothing conclusive, I'll update the PR.

Oh nice, with the `--testcomp` flag I had an instrumentation that changed `&&` and `||` to `logand_` and `logor_`. But I dont see these external function anywhere, so it's probably not implemented :sweat_smile:

File: example/c/dobble.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// An encoding reprensting the problem of finding a suitable
```
Comment (zapashcanon) [inline example/c/dobble.c:1]: s/reprensting/representing/

**Pull request merged**